presto-main-base/src/test/java/com/facebook/presto/sql/planner/optimizations/TestAddExchangesPlansWithFunctions.java

```diff
@@ -0,0 +1,648 @@
+                                                        tableScan("columns", ImmutableMap.of("ordinal_position", "ordinal_position"))))),
+                                anyTree(
+                                        project(ImmutableMap.of("nationkey", expression("nationkey")),
+                                                filter(
+                                                        tableScan("nation", ImmutableMap.of("nationkey", "nationkey"))))))));
+
+        // Regular table with CPP function (should work normally without extra exchange)
+        assertNativeDistributedPlan(
+                "SELECT * FROM nation WHERE foo(nationkey) = 1",
+                anyTree(
+                        exchange(REMOTE_STREAMING, GATHER,
+                                filter("foo(nationkey) = BIGINT'1'",
+                                        tableScan("nation", ImmutableMap.of("nationkey", "nationkey"))))));
+    }
+
+    @Test
+    public void testAggregationsWithMixedFunctions()
+    {
+        // Aggregation with CPP function in GROUP BY
+        assertNativeDistributedPlan(
+                "SELECT DISTINCT foo(ordinal_position) FROM information_schema.columns",
+                anyTree(
+                        project(ImmutableMap.of("foo", expression("foo(ordinal_position)")),
+                                exchange(REMOTE_STREAMING, GATHER,
+                                        tableScan("columns", ImmutableMap.of("ordinal_position", "ordinal_position"))))));
+
+        // Aggregation with Java function in GROUP BY - can be pushed down
+        assertNativeDistributedPlan(
+                "SELECT DISTINCT bar(ordinal_position) FROM information_schema.columns",
+                anyTree(
+                        exchange(REMOTE_STREAMING, GATHER,
+                                project(ImmutableMap.of("bar", expression("bar(ordinal_position)")),
+                                        tableScan("columns", ImmutableMap.of("ordinal_position", "ordinal_position"))))));
+    }
+
+    @Test
+    public void testComplexPredicateWithMultipleFunctions()
+    {
+        // Complex predicate with multiple CPP and Java functions
+        // Since the predicate contains CPP functions (foo, baz), the exchange is inserted before the system table scan
+        // The RemoveRedundantExchanges rule removes the inner exchange that was added by ExtractIneligiblePredicatesFromSystemTableScans
+        assertNativeDistributedPlan(
+                "SELECT * FROM information_schema.columns WHERE (foo(ordinal_position) > 0 AND bar(ordinal_position) < 100) OR baz(ordinal_position) = 50",
+                anyTree(
+                        filter(
+                                exchange(REMOTE_STREAMING, GATHER,
+                                        tableScan("columns", ImmutableMap.of("ordinal_position", "ordinal_position"))))));
+    }
+
+    @Test
+    public void testProjectionWithMixedFunctions()
```

Comment:
How about a `testProjectionAndFilterWithFixedFunctions` ?

Reply:
This is just the projection, there's no filter node here.